### DISCUSSION TITLE
NLMPC seems not to work with soft constraints

### DISCUSSION THREAD

**Discussion by salvatoredeluca** [casadi/casadi#3830] [Q&A]
Hi, I'm trying to implement an MPC algorithm in CasADi to make an underwater vehicle (AUV) move from one point to another while avoiding obstacles. To achieve this, I use a cost function that minimizes the control inputs and the distance from the goal point. I have also added a hard constraint to keep the vehicle away from the obstacles, which worked pretty well!

Now, I want to add a penalty cost (a soft constraint) to encourage the AUV to stay slightly farther from the obstacle rather than closely following the boundary around it. However, the algorithm behaves the same as before, and the path always follows the circle around the obstacle. Any advice?
I dropped both the Matlab Code and the scientific article which implements this type of algorithm.
[NLMPC_AUV.zip](https://github.com/user-attachments/files/16825818/NLMPC_AUV.zip)
[ARTICOLO-OSTACOLI.pdf](https://github.com/user-attachments/files/16825819/ARTICOLO-OSTACOLI.pdf)

**Comment by Wang-Peng-debug** [comment]
Maybe the penalty function is not set well, so try to implement a strong penalty if the AUV gets close to the obstacles or increase the wieight of penalty part.
